Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py

```diff
@@ -202,16 +299,300 @@ def _get_k(
+        distinct_sources = set()
+        for src in self.SOURCES:
+            if src.lower() in owner.get("Source", "").lower():
+                distinct_sources.add(src.lower())
+        return len(distinct_sources)
+
+    def get_min_path_length(self, owner: dict[str, Any]) -> Union[float, int]:
+        """
+        Returns the minimum path length to reach this owner.
+        """
+        min_path_length = float('inf')
+        for src in self._get_sources(owner):
+            src_path_length = 1
+            while src.startswith("Chain: "):
+                src_path_length += 1
+                src = src[len("Chain: "):]
+            if min_path_length is None or src_path_length < min_path_length:
+                min_path_length = src_path_length
+        return min_path_length
+
+    def get_name_similarity_person_asset(self,
+                                         service_identifiers: Iterable[str],
+                                         owner: dict[str, Any]) -> float:
+        """
+        Returns >=1 if there is a blatant match between any `service_identifiers` and `owner`.
+        Returns 0 if there is no match at all.
+        Returns 0 to 1 if there is a potential match.
+        """
+        personal_monikers = [owner.get("Email", ""), owner.get("Name", "")]
+        best_similarity = 0.0
+        for service_id in service_identifiers:
+            similarity = get_name_similarity_index(personal_monikers, service_id)
+            if similarity > best_similarity:
+                best_similarity = similarity
+        return best_similarity
+
+    def get_in_cmdb(self, owner: dict[str, Any]) -> int:
+        """
+        Return 1 if any `owner` is attested in any CMDB; 0 otherwise.
+        """
+        for src in self._get_sources(owner):
+            if "CMDB" in src:
+                return 1
+        return 0
+
+    def get_in_logs(self, owner: dict[str, Any]) -> int:
+        """
+        Return 1 if any `owner` is attested in any logs; 0 otherwise.
+        """
+        for src in self._get_sources(owner):
+            if "Splunk" in src or "log" in src.lower():
+                return 1
+        return 0
+
+    def featurize(self, service_identifiers: Iterable[str], owners: list[dict[str, Any]]) -> np.ndarray:
+        """
+        Generate a featurized numpy array from `service_identifiers` and `owners`.
+        """
+        X = np.zeros((len(owners), len(self.FEATURES)))
+        for sample_idx, owner in enumerate(owners):
+            for feature_idx, (method_name, method) in enumerate(self.FEATURES):
+                try:
+                    if "similarity" in method_name and "person" in method_name and "asset" in method_name:
+                        X[sample_idx, feature_idx] = method(service_identifiers, owner)
+                    else:
+                        X[sample_idx, feature_idx] = method(owner)
```

Comment:
a lot of the stuff that I originally wrote reads like unreviewed prototype code (because it is), but this REALLY reads like unreviewed prototype code -- is there a more robust way to handle it? maybe instead of a single variable `self.FEATURES` there should be 2 variables? other ideas?

(feel free to fix & review the other code too!)